lib/modules/manager/pipenv/extract.ts

```diff
@@ -126,14 +135,34 @@ export async function extractPackageFile(
     return null;
   }
   const res: PackageFileContent = { deps: [] };
-  if (pipfile.source) {
-    res.registryUrls = pipfile.source.map((source) => source.url);
+
+  const sources = pipfile?.source;
+
+  if (sources) {
+    res.registryUrls = sources.map((source) => source.url);
   }
 
-  res.deps = [
-    ...extractFromSection(pipfile, 'packages'),
-    ...extractFromSection(pipfile, 'dev-packages'),
-  ];
+  const categories: [string, Record<string, PipRequirement>][] = [];
+
+  // This should obviously just be a Array.filter() call, but TypeScript doesn't infer the types correctly
+  Object.entries(pipfile).forEach(([category, section]) => {
```

Comment:
we should do the loop once for perf, so please merge with loop below.

use a for-of loop